ledger/complete/mtrie/trie/trie.go

```diff
@@ -313,21 +287,35 @@ func (mt *MTrie) proofs(head *node.Node, paths []ledger.Path, proofs []*ledger.T
 	for _, p := range proofs {
 		p.Steps++
 	}
-	// split paths based on the value of i-th bit (i = trie height - node height)
-	lpaths, lproofs, rpaths, rproofs := utils.SplitTrieProofsByPath(paths, proofs, mt.height-head.Height())
+
+	// partition step to quick sort the paths
+	partitionIndex := utils.SplitTrieProofsByPath(paths, proofs, mt.Height()-head.Height())
+	lpaths, rpaths := paths[:partitionIndex], paths[partitionIndex:]
+	lproofs, rproofs := proofs[:partitionIndex], proofs[partitionIndex:]
+
+	wg := sync.WaitGroup{}
+	parallelRecursionThreshold := 128 // thresold to avoid the parallelization going too deep in the recursion
```

Comment:
I actually implemented this way back on early days but performance wise, the overhead of waitgroups made it worse in my benchmarks.  can we benchmark this part

Reply:
You're right, the timing is worse if the parallelization goes deep in the recursion. However the timing becomes better when I stop the parallelization at some level. Here I tweaked that level till I got the best timing at `128`.  I spoke about it [here](https://axiomzen.slack.com/archives/G01LSR0RQG5/p1614974337014000)